### ISSUE TITLE
Can we inject our ORM for this bencher? Or, what are the criteria?

### ISSUE THREAD

**Issue by mikependon** [FransBouma/RawDataAccessBencher#41]
Hi Frans, I do not know how to contact you, so I hope it is okay to create an issue with correct labels.

We spent too much time developing a lightweight ORM recently and is using this bencher to test the performance against the other.

Thinking if, what's the criteria for our ORM to be a part of this bencher?
- Do we need to be popular or have more users first?

For now, I created a blog how we made our library works as fast-as-possible.
Link: http://codesdirectory.blogspot.com/2018/09/repodb-net-lightweight-orm-library.html

Here is the cloned repository with our ORM.
Link: https://github.com/RepoDb/RawDataAccessBencher

**Comment by mikependon** [comment]
First, I really thank you for your consideration!

We have injected our library in your bencher with the best approach as we can. We have 2 result sets with it, one is via Object-Based (we composed the SQL) and the other one is via RawSql (use composes the SQL). The one you are seeing right now is a RawSql-Based result.

I hope that I am not commanding here, but yes, you have all the full-control to check the performance of our library. You can verify as well the mapped-object results whether data property values are correct after the mapping (to see whether we are cheating or not), but I again, we tried our best that we can do for us to gain that performance.

We also have written the Unit Tests and Integration Tests on this library to make sure that the result is correct and the codes are protected.

Link: https://github.com/RepoDb/RepoDb/tree/master/RepoDb/RepoDb.Tests

**Comment by FransBouma** [comment]
No problem :) 

I've just ran the benchmark on our test system, I skipped NHibernate, Massive and EF 6 as these are dogslow and it's Sunday after all, and these aren't really important for the list. 

```
Results per framework. Values are given as: 'mean (standard deviation)'
==============================================================================
Non-change tracking fetches, set fetches (25 runs), no caching
------------------------------------------------------------------------------
Tortuga Chain, Compiled v1.2.6553.39558                              : 131,29ms (2,35ms)        Enum: 1,32ms (0,07ms)
Handcoded materializer using DbDataReader                            : 132,39ms (12,43ms)       Enum: 1,23ms (0,04ms)
LINQ to DB v2.0.0.0 (v2.0.0) (compiled)                              : 138,85ms (9,42ms)        Enum: 1,26ms (0,12ms)
LINQ to DB v2.0.0.0 (v2.0.0) (normal)                                : 139,05ms (10,27ms)       Enum: 1,20ms (0,04ms)
Handcoded materializer using DbDataReader (GetValues(array), boxing) : 141,01ms (11,74ms)       Enum: 1,44ms (0,09ms)
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco typed view with QuerySpec        : 150,21ms (5,48ms)        Enum: 1,49ms (0,15ms)
Raw DbDataReader materializer using object arrays                    : 150,77ms (12,51ms)       Enum: 3,25ms (0,42ms)
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco with Raw SQL                     : 150,87ms (10,70ms)       Enum: 1,43ms (0,18ms)
PetaPoco Fast v4.0.3                                                 : 150,92ms (10,45ms)       Enum: 1,30ms (0,08ms)
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco typed view with Linq             : 153,41ms (4,88ms)        Enum: 1,27ms (0,12ms)
ServiceStack OrmLite v5.0.0.0 (v5.1.0.0)                             : 163,45ms (9,97ms)        Enum: 1,54ms (0,06ms)
NPoco v3.9.4.0 (v3.9.4.0)                                            : 165,28ms (11,69ms)       Enum: 1,58ms (0,14ms)
PetaPoco v4.0.3                                                      : 169,38ms (10,41ms)       Enum: 1,45ms (0,06ms)
Dapper v1.50.5.0                                                     : 170,53ms (15,76ms)       Enum: 1,59ms (0,12ms)
RepoDb Persistent v1.5.3.0                                           : 178,46ms (9,10ms)        Enum: 1,53ms (0,08ms)
Entity Framework v6.0.0.0 (v6.2.61023.0)                             : 178,56ms (13,39ms)       Enum: 1,62ms (0,03ms)
RepoDb v1.5.3.0                                                      : 186,16ms (10,48ms)       Enum: 1,52ms (0,10ms)
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 196,11ms (12,80ms)       Enum: 1,83ms (0,15ms)
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 197,00ms (13,30ms)       Enum: 1,60ms (0,16ms)
LLBLGen Pro v5.4.0.0 (v5.4.1), DataTable based TypedView             : 239,27ms (19,36ms)       Enum: 7,29ms (1,50ms)
Tortuga Chain v1.2.6553.39558                                        : 286,99ms (5,90ms)        Enum: 1,58ms (0,23ms)

Memory usage, per iteration
------------------------------------------------------------------------------
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco with Raw SQL                     : 15.167 KB (15.532.016 bytes)
PetaPoco Fast v4.0.3                                                 : 15.210 KB (15.575.544 bytes)
Handcoded materializer using DbDataReader                            : 15.286 KB (15.653.432 bytes)
LINQ to DB v2.0.0.0 (v2.0.0) (compiled)                              : 15.976 KB (16.359.864 bytes)
LINQ to DB v2.0.0.0 (v2.0.0) (normal)                                : 15.982 KB (16.366.168 bytes)
Tortuga Chain, Compiled v1.2.6553.39558                              : 16.166 KB (16.554.248 bytes)
PetaPoco v4.0.3                                                      : 21.106 KB (21.613.440 bytes)
Handcoded materializer using DbDataReader (GetValues(array), boxing) : 30.799 KB (31.538.224 bytes)
Dapper v1.50.5.0                                                     : 30.800 KB (31.539.432 bytes)
Raw DbDataReader materializer using object arrays                    : 31.008 KB (31.752.696 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco typed view with QuerySpec        : 31.861 KB (32.625.984 bytes)
Entity Framework v6.0.0.0 (v6.2.61023.0)                             : 32.444 KB (33.222.736 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco typed view with Linq             : 32.477 KB (33.256.592 bytes)
ServiceStack OrmLite v5.0.0.0 (v5.1.0.0)                             : 33.746 KB (34.556.592 bytes)
RepoDb Persistent v1.5.3.0                                           : 33.991 KB (34.807.624 bytes)
RepoDb v1.5.3.0                                                      : 34.052 KB (34.869.544 bytes)
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 34.972 KB (35.811.520 bytes)
NPoco v3.9.4.0 (v3.9.4.0)                                            : 41.015 KB (42.000.200 bytes)
Tortuga Chain v1.2.6553.39558                                        : 43.648 KB (44.696.016 bytes)
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 48.537 KB (49.702.104 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1), DataTable based TypedView             : 55.356 KB (56.684.752 bytes)

Change tracking fetches, set fetches (25 runs), no caching
------------------------------------------------------------------------------
DataTable, using DbDataAdapter                                       : 198,29ms (12,09ms)       Enum: 28,19ms (0,37ms)
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 256,83ms (10,08ms)       Enum: 2,05ms (0,08ms)
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 257,12ms (17,32ms)       Enum: 7,51ms (0,13ms)
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 418,09ms (26,09ms)       Enum: 2,55ms (0,44ms)

Memory usage, per iteration
------------------------------------------------------------------------------
DataTable, using DbDataAdapter                                       : 39.751 KB (40.705.856 bytes)
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 46.613 KB (47.732.400 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 52.744 KB (54.010.592 bytes)
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 89.432 KB (91.578.384 bytes)

Non-change tracking individual fetches (100 elements, 25 runs), no caching
------------------------------------------------------------------------------
RepoDb Persistent v1.5.3.0                                           : 0,40ms (0,03ms) per individual fetch
RepoDb v1.5.3.0                                                      : 0,43ms (0,02ms) per individual fetch
ServiceStack OrmLite v5.0.0.0 (v5.1.0.0)                             : 0,44ms (0,02ms) per individual fetch
Tortuga Chain, Compiled v1.2.6553.39558                              : 0,46ms (0,03ms) per individual fetch
Dapper v1.50.5.0                                                     : 0,48ms (0,06ms) per individual fetch
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco with Raw SQL                     : 0,48ms (0,05ms) per individual fetch
Tortuga Chain v1.2.6553.39558                                        : 0,49ms (0,02ms) per individual fetch
LINQ to DB v2.0.0.0 (v2.0.0) (compiled)                              : 0,50ms (0,02ms) per individual fetch
LINQ to DB v2.0.0.0 (v2.0.0) (normal)                                : 0,54ms (0,02ms) per individual fetch
Handcoded materializer using DbDataReader                            : 0,54ms (0,08ms) per individual fetch
Handcoded materializer using DbDataReader (GetValues(array), boxing) : 0,56ms (0,07ms) per individual fetch
PetaPoco Fast v4.0.3                                                 : 0,58ms (0,03ms) per individual fetch
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco typed view with QuerySpec        : 0,69ms (0,02ms) per individual fetch
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 0,72ms (0,03ms) per individual fetch
LLBLGen Pro v5.4.0.0 (v5.4.1), DataTable based TypedView             : 0,75ms (0,01ms) per individual fetch
Entity Framework v6.0.0.0 (v6.2.61023.0)                             : 0,76ms (0,01ms) per individual fetch
NPoco v3.9.4.0 (v3.9.4.0)                                            : 0,89ms (0,02ms) per individual fetch
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco typed view with Linq             : 1,03ms (0,02ms) per individual fetch
Raw DbDataReader materializer using object arrays                    : 1,10ms (0,08ms) per individual fetch
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 1,51ms (0,01ms) per individual fetch
PetaPoco v4.0.3                                                      : 3,54ms (0,08ms) per individual fetch

Memory usage, per individual element
------------------------------------------------------------------------------
Handcoded materializer using DbDataReader                            : 08 KB (8.192 bytes)
RepoDb Persistent v1.5.3.0                                           : 16 KB (16.384 bytes)
RepoDb v1.5.3.0                                                      : 16 KB (16.384 bytes)
ServiceStack OrmLite v5.0.0.0 (v5.1.0.0)                             : 16 KB (16.384 bytes)
Dapper v1.50.5.0                                                     : 16 KB (16.384 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco with Raw SQL                     : 16 KB (16.384 bytes)
Handcoded materializer using DbDataReader (GetValues(array), boxing) : 16 KB (16.384 bytes)
Tortuga Chain, Compiled v1.2.6553.39558                              : 24 KB (24.576 bytes)
Tortuga Chain v1.2.6553.39558                                        : 32 KB (32.768 bytes)
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 48 KB (49.152 bytes)
PetaPoco Fast v4.0.3                                                 : 56 KB (57.344 bytes)
PetaPoco v4.0.3                                                      : 64 KB (65.536 bytes)
LINQ to DB v2.0.0.0 (v2.0.0) (compiled)                              : 88 KB (90.112 bytes)
LINQ to DB v2.0.0.0 (v2.0.0) (normal)                                : 88 KB (90.112 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco typed view with QuerySpec        : 96 KB (98.304 bytes)
Entity Framework v6.0.0.0 (v6.2.61023.0)                             : 120 KB (122.880 bytes)
NPoco v3.9.4.0 (v3.9.4.0)                                            : 152 KB (155.648 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1), DataTable based TypedView             : 176 KB (180.224 bytes)
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 224 KB (229.376 bytes)
Raw DbDataReader materializer using object arrays                    : 258 KB (264.408 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1), Poco typed view with Linq             : 296 KB (303.104 bytes)

Change tracking individual fetches (100 elements, 25 runs), no caching
------------------------------------------------------------------------------
DataTable, using DbDataAdapter                                       : 0,46ms (0,02ms) per individual fetch
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 0,65ms (0,01ms) per individual fetch
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 0,75ms (0,08ms) per individual fetch
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 1,56ms (0,02ms) per individual fetch

Memory usage, per individual element
------------------------------------------------------------------------------
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 48 KB (49.152 bytes)
DataTable, using DbDataAdapter                                       : 56 KB (57.344 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 72 KB (73.728 bytes)
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 240 KB (245.760 bytes)

Change tracking fetches, eager load fetches, 3-node split graph, 1000 root elements (25 runs), no caching
------------------------------------------------------------------------------
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 58,07ms (1,10ms)
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 59,33ms (3,69ms)
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 106,50ms (1,22ms)

Memory usage, per iteration
------------------------------------------------------------------------------
Linq to Sql v4.0.0.0 (v4.7.3062.0)                                   : 7.251 KB (7.425.376 bytes)
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 18.969 KB (19.424.392 bytes)
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 19.418 KB (19.884.272 bytes)

Async change tracking fetches, eager load fetches, 3-node split graph, 1000 root elements (25 runs), no caching
------------------------------------------------------------------------------
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 61,22ms (1,13ms)
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 152,22ms (2,63ms)

Memory usage, per iteration
------------------------------------------------------------------------------
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 19.502 KB (19.970.248 bytes)
Entity Framework Core v2.0.1.0 (v2.0.1.17303)                        : 26.064 KB (26.690.136 bytes)

Change tracking fetches, set fetches (25 runs), caching
------------------------------------------------------------------------------
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 64,49ms (2,58ms) Enum: 5,82ms (0,34ms)

Memory usage, per iteration
------------------------------------------------------------------------------
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 76.065 KB (77.891.496 bytes)

Change tracking individual fetches (100 elements, 25 runs), caching
------------------------------------------------------------------------------
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 0,14ms (0,00ms) per individual fetch

Memory usage, per individual element
------------------------------------------------------------------------------
LLBLGen Pro v5.4.0.0 (v5.4.1)                                        : 136 KB (139.264 bytes)

Complete. Press enter to exit.
```

The set fetch results don't match the set fetch results in your blogpost, however, but they do match what I expected namely that the handwritten approach is still the fastest (as it should ;), no shame in not beating that, as that's nearly impossible). 

I think your blogpost has the first graph labeled wrong? As your ORM claims to be the fastest in the set fetches, but that's pretty hard to do, as I said (and my tests show you're not, which is OK, the handwritten materializer should be the fastest there in normal situations). The individual fetches are a different story, you are indeed the fastest there, not by a mile, but it's a good result 

Could you create a PR so I can merge it? (and please correct the graph in the blogpost so it properly represents the right fetch ;)). I've updated the benchmarks a bit in a branch with insert tests as well, will push and merge that one in the coming weeks.

**Comment by mikependon** [comment]
Hi Frans,

I am almost crying as I am writing this, and again, I am thanking you for considering this one. This is a big leap for our library for us to be able to continue contribute in the .Net community.

We will assure that we will keep developing this ORM. I will use your result to update my blog, if that is okay. I will create another worksheet on our result-set as well and will do use that graph presentation to update my blog.

Again, I am thanking you for spending time on this one. I will create a PR into your bencher in a while.

Best regards and cheers.

Michael

**Comment by FransBouma** [comment]
:) 

You can use the results, it was a release build run. As your set fetch pipeline is a bit behind the pack, it might be worth looking into compiling a generated expression tree which accepts a datareader and results in a new, materialized poco instance. So generate something like:
`r=>new p() { Field1=r.GetInt32(0), Field2 = r.GetString(1), ... , Fieldn=r.GetInt32(n-1);}` where the type of the get method called of course depends on the type of the field at ordinal n. For fields that are nullable, you have to check for nulls of course, so you then have to generate something like `Field1 = r.IsDBNull(0) ? null : r.GetInt32(0)`, it's up to you of course to determine which fields are null and whether you need to do actual testing.

Linq to DB (and I suspect tortuga does too) check the resultset's schema table prior to consuming the resultset and based on that skip null checks for some fields which are actually nullable (if the meta-data says no nulls are expected, you could skip the check). This is the reason Linq to DB is ahead of the pack (and I suspect tortuga is too, there's no other way to get this close to the handwritten materializer in set fetches). It's unfortunately not reliable in many cases, so for this benchmark it might work, in general cases in might not, hence I skipped implementing it in llblgen pro (and therefore have to settle for a spot in the middle ;))

**Comment by mikependon** [comment]
Thank you fro this Frans. :)

> be worth looking into compiling a generated expression tree which accepts a datareader and results in a new, materialized poco instance

and this.

> check the resultset's schema table prior to consuming the resultset and based on that skip null checks for some fields which are actually nullable (if the meta-data says no nulls are expected, you could skip the check)

I will definitely do both of them. Though the second item requires a first-touch in the DB, but I guess, this is very necessary as we did the same thing in the 'IsIdentity' property definition. Both of this could really improve the performance of RepoDb (not just in Set fetches as I can see it - I guess).

They are both worth looking.

P.S: I will do both of them after introducing 2 important features that is very usable in our company (BulkUpdate and BulkMerge).

**Comment by mikependon** [comment]
Hi Frans, I have asked my friend Rodel to do the PR for this. I already updated the blog with your result at the bottom of it. Thank you very much.

**Comment by rdagumampan** [comment]
PR created. 
Keep calm and let's enjoy the last days of summer :)

**Comment by mikependon** [comment]
Hi @FransBouma, we have made a PR 3 days ago for our library. Is there something that we need to do still? Regards. /Michael

**Comment by FransBouma** [comment]
As you know, Github emails you when a PR is made to your repo. I'm aware but I've a million other things to do and therefore have little time. I'll merge the PR when I have time for it, and at the moment said million other things take a little bit more priority over your PR.

**Comment by mikependon** [comment]
Hi @FransBouma, your suggestion rocks RepoDb. I have tried following your recommendation on using the 'reader.Get<Type>' (this time it is via Expression, and not via IL implementation). On the other hand, I used to pre-cached the 'FieldDefinitions' by having a first touch to skip the null-checking in the Expression Tree.

Below is the new result, though I have failed 2 important integration test for blob data-types (which I need to fix).

Results per framework. Values are given as: 'mean (standard deviation)'
==============================================================================
Non-change tracking fetches, set fetches (25 runs), no caching
------------------------------------------------------------------------------
Tortuga Chain, Compiled v1.2.6553.39558 : 192.26ms (6.36ms)     Enum: 1.96ms (0.09ms)
RepoDb (RawSql) v1.6.0.0                : 195.16ms (7.95ms)     Enum: 2.10ms (0.19ms)
RepoDb (Poco) v1.6.0.0                  : 196.43ms (5.48ms)     Enum: 1.97ms (0.09ms)
PetaPoco Fast v4.0.3                    : 208.94ms (4.95ms)     Enum: 2.24ms (0.22ms)
Dapper v1.50.5.0                        : 236.84ms (16.64ms)    Enum: 2.51ms (0.30ms)
Tortuga Chain v1.2.6553.39558           : 393.91ms (28.42ms)    Enum: 2.18ms (0.33ms)

Memory usage, per iteration
------------------------------------------------------------------------------
RepoDb (Poco) v1.6.0.0                  : 15,168 KB (15,532,904 bytes)
RepoDb (RawSql) v1.6.0.0                : 15,169 KB (15,533,448 bytes)
PetaPoco Fast v4.0.3                    : 15,208 KB (15,574,008 bytes)
Tortuga Chain, Compiled v1.2.6553.39558 : 16,162 KB (16,550,480 bytes)
Dapper v1.50.5.0                        : 30,798 KB (31,537,712 bytes)
Tortuga Chain v1.2.6553.39558           : 43,648 KB (44,696,000 bytes)

Non-change tracking individual fetches (100 elements, 25 runs), no caching
------------------------------------------------------------------------------
Dapper v1.50.5.0                        : 0.32ms (0.02ms) per individual fetch
RepoDb (RawSql) v1.6.0.0                : 0.32ms (0.03ms) per individual fetch
RepoDb (Poco) v1.6.0.0                  : 0.36ms (0.02ms) per individual fetch
Tortuga Chain, Compiled v1.2.6553.39558 : 0.36ms (0.02ms) per individual fetch
Tortuga Chain v1.2.6553.39558           : 0.40ms (0.02ms) per individual fetch
PetaPoco Fast v4.0.3                    : 0.51ms (0.02ms) per individual fetch

Memory usage, per individual element
------------------------------------------------------------------------------
Dapper v1.50.5.0                        : 16 KB (16,384 bytes)
RepoDb (RawSql) v1.6.0.0                : 16 KB (16,384 bytes)
RepoDb (Poco) v1.6.0.0                  : 24 KB (24,576 bytes)
Tortuga Chain, Compiled v1.2.6553.39558 : 24 KB (24,576 bytes)
Tortuga Chain v1.2.6553.39558           : 32 KB (32,768 bytes)
PetaPoco Fast v4.0.3                    : 56 KB (57,344 bytes)

**Comment by mikependon** [comment]
Hi @FransBouma,

Good morning and good day head. I am very interested on this topic and would like to discuss this if you have time soon to answer. With the latest updates we have at v1.6.0, we saw RepoDb is now playing in the top 2-4 ORM in Set-Fetches but eventually degraded a little in the Individual-Fetches. We somehow went from top 1 to top 3 or 4 (just behind Tortuga and Dapper sometimes), luckily sometimes we still ahead of the pack still (but this is now not consistent anymore).

Anyway, the changes made were, we introduced a new compiled-function Expression to process the DataReader convertion to DataEntity. We cached it eventually for the future used. We then use this conversion and forget about the IL.

The codes can be found [here](https://github.com/RepoDb/RepoDb/blob/master/RepoDb/RepoDb/Reflection/FunctionFactory.cs).

Below is the actual Expression Member Assignments for a Class with 8 fields. You will notice that because of the 'FieldDefinition' from the DB, we managed to removed the NULL-Checks at 0, 1, 6, 7.

[0] = {Id = reader.GetInt64(0)}
[1] = {Name = reader.GetString(1)}
[2] = {Address = IIF(reader.IsDBNull(2), default(String), reader.GetString(2))}
[3] = {DateOfBirth = IIF(reader.IsDBNull(3), default(DateTime), reader.GetDateTime(3))}
[4] = {Worth = new Nullable`1(IIF(reader.IsDBNull(4), default(Decimal), reader.GetDecimal(4)))}
[5] = {SsId = new Nullable`1(IIF(reader.IsDBNull(5), default(Guid), reader.GetGuid(5)))}
[6] = {DateUpdated = reader.GetDateTime(6)}
[7] = {DateInserted = reader.GetDateTime(7)}

What I notice here is that, the member assignment is forced. In my previous IL implementation, I have NULL checks and eventually I skip it. Since it is not a forced member assignment, then eventually it is skipping the process of assignment and that really helps to speed up the process. And this could be the reason why we are first before.

My IL Pseudo Code:

```
FUNCTION CONVERT(DBDATAREADER)
{
..... READ FROM DB READER
IF (ISDBNULL(ValueFromDbReader))
BEGIN
      GOTO LABEL_END;
END
..... OTHE OPERATIONS HERE
LABEL_END:
RETURN;
}
```

If we are to convert it to Expression, then it could be looked like this: Let us say if 'Address' and 'LastUpdated' data value is NULL then we will skip it and leaving the property value as its default value.

[0] = {Id = reader.GetInt64(0)}
[1] = {Name = reader.GetString(1)}
[3] = {DateOfBirth = IIF(reader.IsDBNull(3), default(DateTime), reader.GetDateTime(3))}
[4] = {Worth = new Nullable`1(IIF(reader.IsDBNull(4), default(Decimal), reader.GetDecimal(4)))}
[5] = {SsId = new Nullable`1(IIF(reader.IsDBNull(5), default(Guid), reader.GetGuid(5)))}
[7] = {DateInserted = reader.GetDateTime(7)}

But I just could not make it here in Expression approach as I am required to set the property for MemberAssignment (member inits) via NULL-checks (with DbDataReader.IsDBNull method).

**Comment by FransBouma** [comment]
The behavior is different if you skip a set. The CTor of the type could set it to a default value (e.g. "") and the NULL from the DB then doesn't overwrite it in your IL approach, which IMHO isn't correct. I wouldn't worry about being 1st or being 4th by just a few ms. One should start worrying if you're behind the majority by a margin. 

Claiming you're the 'fastest' is a delegate topic, and up to you, but there's always a situation thinkable where you're not. Being 'the fastest' isn't that hard tho: just cut corners and features. The more features the slower you will be. See, my framework pipelines aren't the no.1 as some features (like type conversions on values coming from the DB) require certain methods to be called during materialization, a different materializer approach etc. so performance takes a (minor) hit. I could cut these features, add an unreliable null checker like Linq to DB does and claim top spot. The price paid for that is too high IMHO. 

so in short: don't worry about the forced assignment.
